### PR TITLE
opencv 4.6.0 with related migrations to `jpeg-turbo`, `python@3.10`, and `ffmpeg`

### PR DESCRIPTION
**Pull request by cho-m**
- [ ] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [ ] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [ ] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

Migrations for `vtk`, `opencv`, `gdal`, `mapserver`, `mlt`, `ffms2`, `visp`, `mapnik`

### CODE REVIEW

**Comment by cho-m** [comment]
Audit will at least fail on `mapserver`.

Also want to look into bundled libraries in `opencv` similar to #107133

File: Formula/mapserver.rb [hunk 1]
```diff
@@ -45,44 +45,39 @@ class Mapserver < Formula
...
-      -DPYTHON_EXECUTABLE=#{Formula["python@3.9"].opt_bin/"python3"}
-      -DPHP_EXTENSION_DIR=#{lib}/php/extensions
-      -DCMAKE_INSTALL_RPATH=#{rpath}
-    ]
-
     # Install within our sandbox
-    inreplace "mapscript/python/CMakeLists.txt" do |s|
-      s.gsub! "${PYTHON_LIBRARIES}", "-Wl,-undefined,dynamic_lookup"
-    end
+    inreplace "mapscript/python/CMakeLists.txt", "${PYTHON_LIBRARIES}", "-Wl,-undefined,dynamic_lookup"
```
Comment (cho-m) [inline Formula/mapserver.rb:49]: Not sure about Linux here.

File: Formula/mapserver.rb [hunk 2]
```diff
@@ -45,44 +45,39 @@ class Mapserver < Formula
...
 
-    system "cmake", "-S", ".", "-B", "build", *args, *std_cmake_args
+    ENV.cxx11
```
Comment (cho-m) [inline Formula/mapserver.rb:51]: Not sure if this is needed:
```cmake
set(CMAKE_CXX_STANDARD 11
  CACHE STRING "C++ standard version to use (default is 11)")
```

File: Formula/opencv.rb [hunk 3]
```diff
@@ -27,14 +27,14 @@ class Opencv < Formula
   depends_on "ffmpeg@4"
```
Comment (cho-m) [inline Formula/opencv.rb:27]: Note to check on `ffmpeg@4 --> ffmpeg` too: https://github.com/opencv/opencv/pull/21754

File: Formula/opencv.rb [hunk 4]
```diff
@@ -59,14 +56,21 @@ def install
     # Reset PYTHONPATH, workaround for https://github.com/Homebrew/homebrew-science/pull/4885
     ENV.delete("PYTHONPATH")
 
+    # Remove bundled libraries to make sure formula dependencies are used
+    libdirs = %w[ffmpeg libjasper libjpeg libjpeg-turbo libpng libtiff libwebp openexr openjpeg protobuf tbb zlib]
+    libdirs.each { |l| (buildpath/"3rdparty"/l).rmtree }
```
Comment (cho-m) [inline Formula/opencv.rb:61]: Remaining directories:
```
carotene cpufeatures ippicv ittnotify libtengine libtim-vx openvx quirc
include/opencl
include/vulkan
```

Some aren't used (like `cpufeatures` is Android-only) and Vulkan is disabled by default. Others aren't formulae like Intel libraries.

File: Formula/opencv.rb [hunk 5]
```diff
@@ -92,43 +96,39 @@ def install
...
+      ]
     end
 
-    if Hardware::CPU.intel?
-      args << "-DENABLE_AVX=OFF" << "-DENABLE_AVX2=OFF"
-      args << "-DENABLE_SSE41=OFF" << "-DENABLE_SSE42=OFF" unless MacOS.version.requires_sse42?
+    ENV.runtime_cpu_detection
+    if Hardware::CPU.intel? && build.bottle?
+      args << "-DCPU_BASELINE_REQUIRE=#{MacOS.version.requires_sse42? ? "SSE4_2" : "SSSE3"}"
     end
```
Comment (cho-m) [inline Formula/opencv.rb:121]: May need someone to double check this. Ref: https://github.com/opencv/opencv/wiki/CPU-optimizations-build-options

**Comment by cho-m** [comment]
`caffe`, `siril`, and `visp` need revision bumps.

`mlt` has checksum failure.

File: Formula/mlt.rb [hunk 6]
```diff
@@ -2,8 +2,9 @@ class Mlt < Formula
   desc "Author, manage, and run multitrack audio/video compositions"
   homepage "https://www.mltframework.org/"
   url "https://github.com/mltframework/mlt/releases/download/v7.8.0/mlt-7.8.0.tar.gz"
-  sha256 "4165e62e007e37d65e96517a45817517067897eedef4d83de7208dbd74b1f0f7"
+  sha256 "66606d79f91b400a4d9380a911a5d771a48bd6413447fa2f3713459eba70242d"
```
Comment (cho-m) [inline Formula/mlt.rb:5]: I think this is related to https://github.com/mltframework/mlt/issues/810

File: Formula/visp.rb [hunk 7]
```diff
@@ -53,7 +51,8 @@ def install
              "C Compiler:                  #{ENV.cc}\"")
     end
 
-    system "cmake", ".", "-DBUILD_DEMOS=OFF",
+    system "cmake", ".", "-DUSE_CXX_STANDARD=11",
```
Comment (cho-m) [inline Formula/visp.rb:54]: This actually doesn't work. It looks like some dependency headers need newer standard (e.g. C++14). By not setting this, I see build ends up adding `-std=c++17`.

**Review by carlocab — DISMISSED** [review]
Changes LGTM, but CI seems to be failing for some of these.

**Comment by cho-m** [comment]
Some of CI failures were probably from GitHub package issue a couple days ago.

`mlt` is going to show checksum audit.

Needs a rebase due to `libpq` PR #107797

Will try running CI again sometime today once long timeout is free.

File: Formula/opencv.rb [hunk 8]
```diff
@@ -143,11 +161,11 @@ def install
...
     EOS
-    system ENV.cxx, "-std=c++11", "test.cpp", "-I#{include}/opencv4",
-                    "-o", "test"
-    assert_equal `./test`.strip, version.to_s
+    system ENV.cxx, "-std=c++11", "test.cpp", "-I#{include}/opencv4", "-o", "test"
+    assert_equal shell_output("./test").strip, version.to_s
 
-    output = shell_output(Formula["python@3.9"].opt_bin/"python3 -c 'import cv2; print(cv2.__version__)'")
+    python = Formula["python@3.10"].opt_bin/"python3.10"
+    output = shell_output("#{python} -c 'import cv2; print(cv2.__version__)'")
```
Comment (cho-m) [inline Formula/opencv.rb:168]: Maybe can just use command:
```suggestion
    output = shell_output("python3.10 -c 'import cv2; print(cv2.__version__)'")
```

File: Formula/mapnik.rb [hunk 9]
```diff
@@ -38,8 +47,7 @@ class Mapnik < Formula
 
   def install
     ENV.cxx11
-
-    ENV["PYTHON"] = Formula["python@3.9"].opt_bin/"python3"
+    ENV["PYTHON"] = "python3.9"
```
Comment (cho-m) [inline Formula/mapnik.rb:50]: Python 3.9 due to Scons 3.0.1 on stable. HEAD uses Scons 4.1.0, but Python 3.10 might need Scons 4.4.0. May consider using Homebrew's `scons`, but will leave to explore in separate PR.

File: Formula/opencv.rb [hunk 10]
```diff
@@ -92,43 +116,41 @@ def install
...
 
-    if Hardware::CPU.intel?
-      args << "-DENABLE_AVX=OFF" << "-DENABLE_AVX2=OFF"
-      args << "-DENABLE_SSE41=OFF" << "-DENABLE_SSE42=OFF" unless MacOS.version.requires_sse42?
+    # Ref: https://github.com/opencv/opencv/wiki/CPU-optimizations-build-options
+    ENV.runtime_cpu_detection
+    if Hardware::CPU.intel? && build.bottle?
+      cpu_baseline = MacOS.version.requires_sse42? ? "SSE4_2" : "SSSE3"
+      args += %W[-DCPU_BASELINE=#{cpu_baseline} -DCPU_BASELINE_REQUIRE=#{cpu_baseline}]
     end
```
Comment (cho-m) [inline Formula/opencv.rb:143]: I think for non-bottle builds, it should:
- on Linux, `-march=native` should force CPU detection
   https://github.com/opencv/opencv/blob/4.6.0/cmake/OpenCVCompilerOptimizations.cmake#L162-L166 
- on macOS, the default is auto-detection
   https://github.com/opencv/opencv/blob/4.6.0/cmake/OpenCVCompilerOptimizations.cmake#L319-L321

File: Formula/synfig.rb [hunk 11]
```diff
@@ -5,7 +5,7 @@ class Synfig < Formula
   mirror "https://github.com/synfig/synfig/releases/download/v1.4.2/synfig-1.4.2.tar.gz"
   sha256 "e66688b908ab2f05f87cc5a364f958a1351f101ccab3b3ade33a926453002f4e"
   license "GPL-3.0-or-later"
-  revision 1
+  revision 2
```
Comment (cho-m) [inline Formula/synfig.rb:8]: Probably needs https://github.com/synfig/synfig/commit/f684b24f0db31ab8ea7aadc417fc23e3084b4138.

Will try locally to check if patch applies and there are no other issues.

**Review by carlocab — DISMISSED** [review]

**Review by carlocab — DISMISSED** [review]
Only `audit` failures for `mlt`.

**Review by carlocab — APPROVED** [review]
Forgot to pass `--no-autosquash`.

File: Formula/opencv.rb (hunk 10 above)
Comment (carlocab) [reply Formula/opencv.rb:143]: This might need adjusting for source builds, then, since `ENV.runtime_cpu_detection` will skip passing any `-march` flags.

Comment (cho-m) [reply Formula/opencv.rb:143]: In that case, Linux install will probably be similar between bottles and source build as SSE3 is default baseline which should give same result as SSSE3.

There is a `CPU_BASELINE=NATIVE` which I didn't see any documentation on but may provide a more tuned variant.

Comment (carlocab) [reply Formula/opencv.rb:143]: I'm ok to set `CPU_BASELINE=NATIVE` if we're not building bottles as long as it doesn't lead to dependent weirdness as with the Facebook formulae.
